paperqa/agents/models.py

```diff
@@ -0,0 +1,520 @@
+from openai import AsyncOpenAI
+from pydantic import (
+    BaseModel,
+    ConfigDict,
+    Field,
+    PrivateAttr,
+    ValidationInfo,
+    computed_field,
+    field_validator,
+    model_validator,
+)
+from typing_extensions import Protocol
+
+from .. import (
+    Answer,
+    OpenAILLMModel,
+    PromptCollection,
+    llm_model_factory,
+)
+from ..utils import hexdigest
+from ..version import __version__
+from .prompts import STATIC_PROMPTS
+
+logger = logging.getLogger(__name__)
+
+
+class SupportsPickle(Protocol):
+    """Type protocol for typing any object that supports pickling."""
+
+    def __reduce__(self) -> str | tuple[Any, ...]: ...
+    def __getstate__(self) -> object: ...
+    def __setstate__(self, state: object) -> None: ...
+
+
+class AgentStatus(str, Enum):
+    # FAIL - no answer could be generated
+    FAIL = "fail"
+    # SUCCESS - answer was generated
+    SUCCESS = "success"
+    # TIMEOUT - agent took too long, but an answer was generated
+    TIMEOUT = "timeout"
+    # UNSURE - the agent was unsure, but an answer is present
+    UNSURE = "unsure"
+
+
+class AgentPromptCollection(BaseModel):
+    """Configuration for the agent."""
+
+    model_config = ConfigDict(frozen=True, arbitrary_types_allowed=True)
+
+    agent_system_prompt: str | None = Field(
+        # Matching https://github.com/langchain-ai/langchain/blob/langchain%3D%3D0.2.3/libs/langchain/langchain/agents/openai_functions_agent/base.py#L213-L215
+        default="You are a helpful AI assistant.",
+        description="Optional system prompt message to precede the below agent_prompt.",
+    )
+
+    # TODO: make this prompt more minimalist, instead improving tool descriptions so
+    # how to use them together can be intuited, and exposing them for configuration
+    agent_prompt: str = (
+        "Answer question: {question}"
+        "\n\nSearch for papers, gather evidence, collect papers cited in evidence then re-gather evidence, and answer."
+        " Gathering evidence will do nothing if you have not done a new search or collected new papers."
+        " If you do not have enough evidence to generate a good answer, you can:"
+        "\n- Search for more papers (preferred)"
+        "\n- Collect papers cited by previous evidence (preferred)"
+        "\n- Gather more evidence using a different phrase"
+        "\nIf you search for more papers or collect new papers cited by previous evidence,"
+        " remember to gather evidence again."
+        " Once you have five or more pieces of evidence from multiple sources, or you have tried a few times, "
+        "call {gen_answer_tool_name} tool. The {gen_answer_tool_name} tool output is visible to the user, "
+        "so you do not need to restate the answer and can simply terminate if the answer looks sufficient. "
+        "The current status of evidence/papers/cost is {status}"
+    )
+    paper_directory: str | os.PathLike = Field(
+        default=Path("."),
```

Comment:
What does `"."` mean? Does it mean `paperqa/agents`, or the caller's directory?

Do you mind making the default be not a relative path?

Reply:
Callers current directory -- it gets turned into an absolute path for you in the function. This is a little complicated because as a CLI, it needs to be whatever directory the user is in OR an absolute path if specified. This way it supports both, and it reflects the default behavior, i.e. you move to a directory and then it runs there.

Reply:
Can you add a comment or description just documenting this then?

Reply:
actually can just use `Path.cwd()`